### PR TITLE
Feat: Network upgrade

### PR DESCRIPTION
**Pull request by tuul-wq**
- Moved network services to `shared/api` according to FSD
- Removed `chainSpecService`
- Cleaned up `metadataService`
- Network domain utils
- New tests / updated tests
- **Fixed: Save metadata**
- Update provider's metadata on new metadata via `provider.updateMetadata`

### CODE REVIEW

File: src/renderer/entities/network/model/network-model.ts [hunk 1]
```diff
@@ -4,20 +4,10 @@ import { ProviderInterface } from '@polkadot/rpc-provider/types';
...
 import { cloneDeep, keyBy } from 'lodash';
 
-import { Metadata, ProviderType, chainsService, isEnabled, networkService } from '../lib';
-import { Chain, ChainId, Connection, ConnectionStatus, ConnectionType, RpcNode } from '@shared/core';
-import { useMetadata } from '../lib/metadataService';
+import { Chain, ChainId, Connection, ConnectionStatus, ConnectionType, RpcNode, Metadata, NoID } from '@shared/core';
+import { ProviderType, chainsService, networkService, metadataService } from '@shared/api/network';
 import { storageService } from '@shared/api/storage';
-
-const chains = chainsService.getChainsMap({ sort: true });
```
Comment (tuul-wq) [inline src/renderer/entities/network/model/network-model.ts:12]: Now we get `chains` as it's supposed to be via effect

File: src/renderer/entities/network/model/network-model.ts [hunk 2]
```diff
@@ -164,34 +133,36 @@ const reconnectFx = createEffect(async ({ api }: ReconnectParams): Promise<void>
...
-      console.info('🔴 provider error ==> ', chainId);
-
-      boundFailed(chainId);
-    },
-  );
-});
+    return networkService.createProvider(
+      chainId,
+      providerType,
+      { nodes, metadata: metadata?.metadata },
```
Comment (tuul-wq) [inline src/renderer/entities/network/model/network-model.ts:148]: `HexString` was enough

File: src/renderer/shared/api/network/service/networkService.ts [hunk 3]
```diff
@@ -19,66 +15,57 @@ export const networkService = {
...
+  const provider = creatorFn[providerType]();
 
   if (!provider) {
     throw new Error('Provider not found');
   }
 
-  provider.on('connected', onConnected || noop);
-  provider.on('disconnected', onDisconnected || noop);
-  provider.on('error', onError || noop);
+  provider.on('connected', listeners.onConnected);
```
Comment (tuul-wq) [inline src/renderer/shared/api/network/service/networkService.ts:48]: No point in provider if we don't assign listeners

File: src/renderer/widgets/WalletDetails/ui/wallets/MultisigWalletDetails.tsx [hunk 4]
```diff
@@ -35,7 +36,9 @@ export const MultisigWalletDetails = ({ wallet, account, signatoryWallets, signa
   const [isConfirmForgetOpen, toggleConfirmForget] = useToggle();
 
   const chains = useMemo(() => {
-    return chainsService.getChainsData({ sort: true }).filter((chain) => isMultisigAvailable(chain.options));
+    return chainsService
+      .getChainsData({ sort: true })
```
Comment (Asmadek) [inline src/renderer/widgets/WalletDetails/ui/wallets/MultisigWalletDetails.tsx:40]: maybe use `networkModel.$chains` here too?

File: src/renderer/entities/network/model/network-model.ts [hunk 5]
```diff
@@ -221,26 +213,26 @@ sample({
   source: {
     chains: $chains,
     connections: $connections,
+    metadata: $metadata,
```
Comment (Asmadek) [inline src/renderer/entities/network/model/network-model.ts:216]: why do you need `metadata` here and `$metadata` at all? You use `networkUtils.getNewestMetadata` to get metadata, not `$metadata`

File: src/renderer/entities/network/model/network-model.ts [hunk 6]
```diff
@@ -322,6 +332,9 @@ sample({
   target: $metadataSubscriptions,
 });
 
+// TODO: matadata save was lost, turn it back task: https://app.clickup.com/t/8693kq4b8
+// sample({ clock: requestMetadataFx.doneData, target: saveMetadataFx });
```
Comment (Asmadek) [inline src/renderer/entities/network/model/network-model.ts:336]: Maybe uncomment this code and `saveMetadataFx` and not create new task for it?

**Review by Asmadek — COMMENTED** [review]
Some questions

File: src/renderer/entities/staking/lib/staking-utils.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import type { ChainId } from '@shared/core';
+
+export const stakingUtils = {
+  isKusamaChainId,
+};
+
+function isKusamaChainId(chainId: ChainId): boolean {
```
Comment (egor0798) [inline src/renderer/entities/staking/lib/staking-utils.ts:7]: Why do we need that?

Comment (tuul-wq) [reply src/renderer/entities/staking/lib/staking-utils.ts:7]: It's used here, temporary fix for Kusama's nominators
```javascript
function getDefaultValidatorsAmount(api: ApiPromise): number {
  if (stakingUtils.isKusamaChainId(api.genesisHash.toHex())) return KUSAMA_MAX_NOMINATORS;

  return DEFAULT_MAX_NOMINATORS;
}
```

File: src/renderer/entities/network/model/network-model.ts (hunk 5 above)
Comment (tuul-wq) [reply src/renderer/entities/network/model/network-model.ts:216]: `networkUtils.getNewestMetadata` works with existing metadata
At first we load all metadata from db
Then pass the newest version to `createCachedProvider`
Previously we requested metadata from db during `send` method inside `createCachedProvider` which is bad

**Comment by tuul-wq** [comment]
Task linked: [CU-8693kq4b8 Turn on Metadata saves](https://app.clickup.com/t/8693kq4b8)

File: src/renderer/shared/api/network/lib/utils.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+export function isKusama(chainName: string): boolean {
+  return chainName === 'Kusama';
+}
+
+export function isTestnet(chainOptions?: ChainOptions[]): boolean {
+  return Boolean(chainOptions?.includes('testnet'));
+}
+
+export function isNameWithNumber(chainName: string): boolean {
+  return /^[0-9]+/.test(chainName);
```
Comment (egor0798) [inline src/renderer/shared/api/network/lib/utils.ts:28]: regexp checks that name starts with number so maybe name the function `isNameStartsWithNumber`?

**Review by egor0798 — APPROVED** [review]

**Review by Asmadek — APPROVED** [review]

**Pull request merged**